website/blog/2022-08-17-managing-surrogate-keys-in-dbt.md

```diff
@@ -0,0 +1,228 @@
+
+### How were surrogate keys managed in the past?
+
+Before the advent of the analytical warehouse tools we use today, the data warehouse architecture had a few key constraints that led to the rise of the Kimball-style warehouse with a snowflake schema. This was because storage was expensive — it was more efficient to store data as few times as possible, and rely on joins to connect data tog   ether when a report required it. And to make those joins efficient, it became standard practice to use **monotonically increasing integer surrogate keys (MIISKs)**, a fancy way to say “count each record starting at one” so that your data model would look something like this (you are a cheesemonger):
+
+| product_id | product_name | created_by | created_at |
+| --- | --- | --- | --- |
+| 1 | mozzarella | 23 | 0001-05-05 |
+| 2 | cheddar | 24 | 1150-02-03 |
+| 3 | gruyere | 25 | 1655-04-03 |
+
+| order_line_id | order_id | product_id | amount | created_at |
+| --- | --- | --- | --- | --- |
+| 1 | 1 | 3 | 40 | 2022-07-01 |
+| 2 | 2 | 1 | 50 | 2022-07-05 |
+| 3 | 3 | 1 | 10 | 2022-07-07 |
+| 4 | 3 | 2 | 30 | 2022-07-07 |
+
+| order_id | customer_id | created_at |
+| --- | --- | --- |
+| 1 | 5 | 2022-07-01 |
+| 2 | 8 | 2022-07-05 |
+| 3 | 10 | 2022-07-07 |
+
+There are some clear benefits here!
+
+- There are clear, intuitive relationships between these entities!
+- The fact that the keys here are small integers, the database can a) not worry about storage costs for this data b) index this field easily, making joins quick and efficient.
+
+However, there are also some clear maintenance issues here. Making updates to, say, your products table will require some careful surgical work to ensure the association of cheddar to id 2 is never accidentally changed. You may have heard of the phrase “load your dims before your facts” — this refers to the careful work required to maintain this referential integrity. Additionally, you need to know about the *exact state of the data* before making any updates. This data is **stateful**, making it rigid and more difficult to work with should there be any losses to this data. Imagine trying to rebuild these relationships from scratch!
+
+## MIISKs in dbt
+
+If this is your preferred modeling approach, dbt can absolutely support this workflow! This will likely require you to take advantage of built-in warehouse functionality to generate these MIISKs — in Snowflake, we can use [sequences](https://docs.snowflake.com/en/user-guide/querying-sequences.html), which are objects built exactly for this purpose. We’ll use Snowflake as the example here, but this approach can likely be adapted for other warehouses as well. 
+
+### Creating and Maintaining Sequences
```

Comment:
### Creating and maintaining sequences

Idk why won't let me make an in-line comment here

Reply:
but the suggestion was to change the capitalization to: Creating and maintaining sequences

Reply:
will update!